Add client codegen guide to docs (#1586)
awslabs/smithy: docs/source-2.0/guides/generating-code/generating-a-client.rst

.. _generating-a-client:

===================
Generating a client
===================

Smithy code generators are implemented as :ref:`smithy-build plugins <plugins>`,

> running when the Smithy model is built. The plugins generate the equivalent of

Reviewer: ```suggestion
that run when the Smithy model is built. The plugins generate the equivalent of
```

> Smithy code generators are implemented as :ref:`smithy-build plugins <plugins>`,
> running when the Smithy model is built. The plugins generate the equivalent of

Reviewer: ```suggestion
Smithy code generators are implemented as :ref:`smithy-build plugins <plugins>`
that run when the Smithy model is built. The plugins generate the equivalent of
```

> packages in their respective languages, configured to be used as such by the
> languages' tooling.

Reviewer: ```suggestion
packages in their respective languages, configured to be used by idiomatic language
tooling.
```

Add the Codegen Plugin
======================

TypeScript clients are generated by the ``typescript-codegen`` plugin. Configure
the plugin by adding it to ``smithy-build.json``:

.. code-block:: json

    {
        "version": "1.0",
        "plugins": {
            "typescript-codegen": {
                "package": "@weather-service/client",
                "packageVersion": "0.0.1"
            }
        }
    }

In this case, we've configured the ``typescript-codegen`` plugin to generate a package
named ``@weather-service/client`` with version ``0.0.1``.

Next, add a build-time dependency on the code generator:

.. code-block:: kotlin

    buildscript {
        repositories {
            mavenCentral()
        }
        dependencies {

>             classpath("software.amazon.smithy.typescript:smithy-aws-typescript-codegen:0.12.0")

Author: Nice call. It's mentioned in the previous page that the code generator must know how to generate the protocol, so I'll link it back to that. I think that saying the code generator has the restJson1 protocol generator might be enough to explain why we depend on it.

>             classpath("software.amazon.smithy.typescript:smithy-aws-typescript-codegen:0.12.0")

Author: Yea this is a minor rough edge. We can absolutely template it, same way we template the smithy version, which has value in and of itself just for reducing duplication. Some time down the road we will have to figure out how we can keep the version in sync with the actual most recent version

>             classpath("software.amazon.smithy.typescript:smithy-aws-typescript-codegen:0.12.0")

Author: Additionally, if we have a directory containing the complete examples, we will want the templated version to be the same in that directory as well. I think we could use Gradle in some way to do that.

        }
    }

> Now run ``gradle build`` to build the model and generate the code. The Typescript

Reviewer: ```suggestion
Now run ``gradle build`` to build the model and generate the code. The TypeScript
```

package is written to the ``typescript-codegen`` directory::

    .
    ├── build
    │   ├── smithyprojections
    │   │   └── weather-service
    │   │       └── source
    │   │           ├── build-info/
    │   │           ├── model/
    │   │           ├── sources/
    │   │           └── typescript-codegen
    │   │               ├── LICENSE
    │   │               ├── package.json
    │   │               ├── src/
    │   │               ├── tsconfig.cjs.json
    │   │               ├── tsconfig.es.json
    │   │               ├── tsconfig.json
    │   │               ├── tsconfig.types.json
    │   │               └── typedoc.json
    │   └── tmp
    ├── build.gradle.kts
    ├── model
    │   └── weather.smithy
    └── smithy-build.json

> 

Reviewer: nit: extra new line

    .
    └── smithy
        ├── build
        ├── build.gradle.kts
        ├── model
        └── smithy-build.json

Next, create a ``package.json`` in the root of the project with the following
contents:

.. code-block:: json

    {
      "name": "weather-service",
      "scripts": {
        "generate": "cd smithy && gradle clean build",
        "build": "yarn workspace @weather-service/client build",
      },
      "dependencies": {
        "@weather-service/client": "0.0.1"
      },
      "private": true,
      "workspaces": [
        "smithy/build/smithyprojections/smithy/client/typescript-codegen"
      ]
    }

A few things to note:

* The path under ``workspaces`` is the path to the root of the generated
  TypeScript package.
* The ``build`` script compiles the generated TypeScript package,
  referred to by the name specified in the ``typescript-codegen`` plugin
  configuration in ``smithy-build.json``.
* A ``generate`` script which builds the model, re-generating the code.
* A dependency has been added on the generated TypeScript package, using the
  name and version specified in the ``typescript-codegen`` plugin configuration
  in ``smithy-build.json``

> Finally, create an ``app.ts`` file to use the client:

Reviewer: The series of commands to build the workspace is not clear (e.g. is `generate` or `build` called first)?
